practice/7_concurrency/task1_fibonacci/template.py

```diff
@@ -12,23 +14,29 @@
 # Overwrite the maximum limit for integer string conversion 
 sys.set_int_max_str_digits(9999999)
 
-def fib(n: int) -> int:
-    """Calculate a value in the Fibonacci sequence by ordinal number"""
-    f0, f1 = 0, 1
-    if n==0: return 0
-    for _ in range(n-1):
-        f0, f1 = f1, f0 + f1
-    return f1
-
-def write_to_file(el: int) -> None:
+# def fib(n: int) -> int:
+#     """Calculate a value in the Fibonacci sequence by ordinal number"""
+#     f0, f1 = 0, 1
+#     if n==0: return 0
+#     for _ in range(n-1):
+#         f0, f1 = f1, f0 + f1
+#     return f1
+
+def fib(n: int, memo: dict) -> int:
+    if n in memo:
+        return memo[n]
+    memo[n] = fib(n-1,memo) + fib(n-2,memo)
```

Comment:
This is a top-down recursion, i.e. to find the result of fib(10000), we have to find fib(9999) and fib(9998) first and so on. So the recursion stack becomes huge and leads to this error. 

Should we use cache? if I've found the value of fib(10000) once, and now I'm asked to find fib(9000), I should already know the answer if I used the cache.

Try to think of pros and cons of different scenarios, e.g. what if we use iterative approach and also cache? or may be it's better to find fib(max(arr)) iteratively and cache all the results, so we actually don't write to the cache and only read from it?